bson/src/main/org/bson/Document.java

```diff
@@ -259,6 +261,63 @@ public Date getDate(final Object key) {
         return (Date) get(key);
     }
 
+    /**
+     * Gets the list value of the given key, casting the list elements to the given {@code Class<T>}.  This is useful to avoid having
+     * casts in client code, though the effect is the same.
+     *
+     * @param key   the key
+     * @param clazz the non-null class to cast the list value to
+     * @param <T>   the type of the class
+     * @return the list value of the given key, or null if the instance does not contain this key.
+     * @throws ClassCastException if the elements in the list value of the given key is not of type T or the value is not a list
+     */
+    public <T> List<T> getList(final Object key, final Class<T> clazz) {
+        notNull("clazz", clazz);
+        return constructValuesList(key, clazz, null);
+    }
+
+    /**
+     * Gets the list value of the given key, casting the list elements to {@code Class<T>} or returning the default list value if null.
+     * This is useful to avoid having casts in client code, though the effect is the same.
+     *
+     * @param key   the key
+     * @param clazz the non-null class to cast the list value to
+     * @param defaultValue what to return if the value is null
+     * @param <T>   the type of the class
+     * @return the list value of the given key, or the default list value if the instance does not contain this key.
+     * @throws ClassCastException if the value of the given key is not of type T
+     * @since 3.5
+     */
+    public <T> List<T> getList(final Object key, final Class<T> clazz, final List<T> defaultValue) {
+        notNull("defaultValue", defaultValue);
+        notNull("clazz", clazz);
+        return constructValuesList(key, clazz, defaultValue);
+    }
+
+    /**
+     * Construct the list of values for the specified key, or return the default value if the value is null.
+     * @param key    the key
+     * @param clazz  the non-null class to cast the list value to
+     * @param defaultValue  what to return if value is null
+     * @param <T>    the type of the class
+     * @return the list value of the given key, or the default list value if the instance does not contain the key.
+     * @throws ClassCastException if the value of the given type is not of type T or value is not a list
+     */
+    @SuppressWarnings("unchecked")
+    private <T> List<T> constructValuesList(final Object key, final Class<T> clazz, final List<T> defaultValue) {
+        List<?> value = get(key, List.class);
+        if (value == null) {
+            return defaultValue;
+        }
+
+        for (Object item : value) {
+            if (!clazz.isInstance(item)) {
```

Comment:
I would loosen this to: `clazz.isAssignableFrom(item.getClass())`

Reply:
And add a test, i.e pass in `Class<Number` and decode a list of mixed Integer and Long values

Reply:
I have added the request test.